---
title: 'Cottrazm: Construct Tumor Transition Zone Microenvironment'
author: "Zhenzhen Xun, Xinyu Ding, Youqiong Ye#"
date: "2022-1-23"
output:
  pdf_document:
    latex_engine: xelatex
  html_document:
    df_print: paged
    toc: yes
  world_document:
    toc: yes
documentclass: ctexart  
bibliography: MyCollection.bib
csl: nature.csl

vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1. Introduction
  Cottrazm is an R package to construct tumor spatial microenvironment (TSME) of spatial transcriptomic (ST) data.
  
  Conttrazm adjusted gene expression with morphological information [@Pham2020], and based on morphological adjusted gene expression,  Cottrazm applied InferCNV [@Tirosh2016] to recognize malignant spots (Mal). Further, Cottrazm arranged spatial spots on hexagonal systems, extrapolated layer by layer from core spots of Mal and determined the identity of spots according to UMAP distance to tumor centroid as Mal or boundary spot (Bdy). When all neighbors of Mal spots are not tumor tissue, the extrapolation will be completed, the remained spots are labeled as non-malignant spots (nMal).
  Next based on corresponding single cell RNA sequencing (scRNAseq) data of tumor, Cottrazm generated signature score matrix from each cell type in scRNAseq, enrichment score matrix of cell types from scRNAseq in each spot by parametric analysis of gene set enrichment (PAGE) analysis, and topic combined the KNN clusters and location information (Mal, Bdy, and nMal). Then, cell types for each topic were determined based on signature score and enrichment score. Cottrazm calculated the possible proportion of each cell type for each spatial spot by dampened weighted least squares (DWLS), a kind of deconvolution algorithm [@Dong2021], to further explore the character of boundary spots as well as malignant and non-malignant spots. 
  Then according to deconvolution result, scRNAseq expression data, and ST expression data, Cottrazm reconstructed the cell type specific gene expression profile (GEP) at sub-spot level, to deeper exploring TSME. Features were weighted in each cell type according to the feature's contribution in each cell type in scRNAseq reference, then feature expression of sub-spot with a certain cell type was calculated by the cell proportion from deconvolution results and feature weight. Cottrazm can be followed by further analysis, including sub-spot GEP analysis, cell-cell interaction, identification of potential targets of tumor boundary. 
  
  The full Cottrazm documentation is available in the package. To reach the user's guide, install the Cottrazm package and load it into an R session by library (Cottrazm). And they can get help by help (BoundaryDefine), help (SpatialDecon), and help (SpatialRecon) to see documentation of the main function. Also description of each function is including in docs/Cottrazm_0.1.1.pdf.
  
# 2. Installation and requirement
Before installing cottrazm, please properly create conda environment install all dependencies 
```{bash}
# conda create -n TumorBoundary python=3.8
# conda activate TumorBoundary 
# pip install -U stlearn 
```

```{r, message=FALSE,warning=FALSE}
library(Seurat)
library(magrittr)
library(dplyr)
library(Matrix)
library(ggplot2)
library(stringr)
library(RColorBrewer)
library(patchwork)
library(ggtree)
library(BiocGenerics)
library(readr)
library(rtracklayer)
library(infercnv)
library(phylogram)
library(utils)
library(dendextend)
library(assertthat)
library(reticulate)
library(openxlsx)
library(scatterpie)
library(cowplot)
library(stats)
library(quadprog)
library(data.table)
library(Rfast)
library(ggrepel)
library(tibble)
library(clusterProfiler)
library(utils)
library(org.Hs.eg.db)
```

Installation of Cottrazm
You can install Cottrazm.tar.gz from local path or directly from github (https://github.com/Yelab2020/Cottrazm).
```{r eval=FALSE}
#from local path
devtools::install_github('Yelab2020/Cottrazm')
#from github
install.packages("Cottrazm.tar.gz", repos = NULL, type = "source")
library(Cottrazm)
```

# 3. functional modules
## 3.1 Delineation of tumor boundary
### 3.1.1 Read, preprocessed and quality control of tumor ST data
We firstly read Spaceranger result, then we examined the quality control result with parameters: nCount_Spatial, nFeature_Spatial and percentage of mitochondria genes. The feature plot and .xlsx file of quality control were saved.
```{r eval = FALSE}
print('STPreProcess')
InDir = paste(system.file("extdata/outs",package = "Cottrazm"),"/",sep = "")
Sample = "CRC1"
OutDir = paste(getwd(),"/",Sample,"/",sep = "")
TumorST <-
  STPreProcess(
    InDir = InDir,
    OutDir = OutDir,
    Sample = Sample
    )
```

### 3.1.2 Morphological adjusted cluster determination
We firstly use morphological information to adjust spatial feature expression, then the adjusted expression data was used to cluster spatial spots. A list of genes represent lymphocytes was used to score spatial spots to recognize normal spatial cluster, the cluster with the highest score of lymphocytes' features was defined as normal cluster. Then we used the barcode and the result of morphological adjusted cluster to generate cell annotation file for future use in InferCNV.
```{r eval = FALSE}
print('STModiCluster')
res = 1.5
TumorST <-
  STModiCluster(
    InDir = InDir,
    Sample = Sample,
    OutDir = OutDir,
    TumorST = TumorST,
    res = res
  )
```

### 3.1.3 Run InferCNV on ST data
We firstly use clusters with the highest score of lymphocytes' features as reference, because the CNV copy number of lymphocytes were relatively stable. Then we run InferCNV on ST data. You can select non-adjusted gene expression assay when tissue was properly permeabilized otherwise we recommend use morphological adjusted gene expression assay. When running InferCNV, the parameter "analysis_mode" was set as random_trees, the observation spots will be divided to 8 clusters, it will take longer time, but malignant and non-malignant spots will be better separated.

```{r eval = FALSE}
print('STCNV' )
STInferCNV <-
  STCNV(
    TumorST = TumorST,
    OutDir  = OutDir,
    assay = "Spatial"
    )
```

Then we score ST data based on result of InferCNV, visualize CNV scores of each CNV labels defined by random_tree model. Next according to CNV scores, CNV labels, and morphology information, you need to assign malignant labels.
```{r eval = FALSE}
print('STCNVScore')

TumorST <-
  STCNVScore(
    TumorST = TumorST,
    assay = "Spatial",
    Sample = Sample,
    OutDir = OutDir
    )
```


### 3.1.4 Define of tumor boundary
According to plots above, CNV labels 1 and 2 had the highest CNV scores and share similar morphological state, so we recognize these labels as putative malignant labels.
Next, we start define the Boundary of putative malignant labels, the processed plots and data were saved in OutDir:
```{r eval = FALSE}
TumorSTn <-
  BoundaryDefine(
    TumorST = TumorST,
    MalLabel = c(7,8),
    OutDir = OutDir,
    Sample = Sample
  )
```

After defined the Boundary of malignant spots, we need to project the malignant spots (Mal), Boundary spots (Bdy) and non-malignant spots (nMal) to spatial HE/IF staining plot. Finally, we generated and saved a boundary defined Seurat object for future analysis.
```{r eval = FALSE}
TumorST <-
  BoundaryPlot(
    TumorSTn = TumorSTn,
    TumorST = TumorST,
    OutDir = OutDir,
    Sample = Sample
  )
```

```{r echo=FALSE}
TumorST <- readr::read_rds('/work/xzz123/Project/2021/TumorBoundary/package/CRC1/CRC1_BoundaryDefine.rds.gz')
Seurat::SpatialDimPlot(TumorST,group.by = 'Location',cols = c("#CB181D", "#1f78b4", "#fdb462"))
```

## 3.2 Deconvolution for each spot
To better understand the cell composition of tumor boundary as well as malignant and other non-malignant spots, we combined tumor ST data and tumor scRNAseq data and used deconvolution method to estimate the proportion of cell types in scRNAseq data infiltrated in each spatial spot.

### 3.2.1 Preporcess of scRNAseq data
Firstly, we need to generate significant gene expression file (sig_exp) and a list of markers for each scRNAseq cell type (clustermarkers_list).
```{r eval = FALSE}
#clustermarkers_list
sc_obj <- readr::read_rds("YourPath/scRNAseqRef.rds.gz")
clustermarkers <-
  Seurat::FindAllMarkers(object = sc_obj,
                         logfc.threshold = 0.25,
                         only.pos = TRUE)

clustermarkers_list <- split(clustermarkers, clustermarkers$cluster)
clustermarkers_list <-
  lapply(names(clustermarkers_list), function(cluster) {
    sub_markers <- clustermarkers_list[[cluster]]$gene
  })
names(clustermarkers_list) <-
  names(split(clustermarkers, clustermarkers$cluster))

#sig_exp
sig_exp <-
  get_sig_exp(
    se.obj = sc_obj,
    DefineTypes = "MajorTypes",
    sig_scran = unique(unlist(clustermarkers_list))
  )

#In this vignette we read processed clustermarkers_list and sig_exp
clustermarkers_list <- readr::read_rds(system.file("inst/extdata/clustermarkers_list.rds.gz",package = "Cottrazm"))
sig_exp <- readr::read_rds(system.file("inst/extdata/sig_exp.rds.gz",package = "Cottrazm"))
```

### 3.2.2  Enrichment analysis of ST data
Then, we preprocessed ST data to get signature score matrix and applied PAGE enrichment analysis on ST data to get the enrichment score matirx of cell types in scRNAseq data of each spatial spot.
```{r eval = FALSE}
#Processed ST data
TumorST <- NormalizeData(TumorST, assay = "Spatial")
TumorST@meta.data$Decon_topics <-
  paste(TumorST@meta.data$Location,
        TumorST@meta.data$seurat_clusters,
        sep = "_")
expr_values = as.matrix(TumorST@assays$Spatial@data) #ST expr log
nolog_expr = 2 ^ (expr_values) - 1 #ST expr nolog
meta_data <-
  TumorST@meta.data[, c("nCount_Spatial", "Decon_topics", "Location")]

#Signature score
for (cluster in names(clustermarkers_list)) {
  cluster_markers = clustermarkers_list[[cluster]][1:25]
  cluster_score <-
    apply(TumorST@assays$Spatial@data[rownames(TumorST@assays$Spatial@data) %in% cluster_markers, ], 2, mean)
  meta_data <- cbind(meta_data, cluster_score)
}
colnames(meta_data) <-
  c("nCount_Spatial",
    "Decon_topics",
    "Location",
    names(clustermarkers_list))

#filter st and sc feature
intersect_gene = intersect(rownames(sig_exp), rownames(nolog_expr))
filter_sig = sig_exp[intersect_gene, ]
filter_expr = nolog_expr[intersect_gene, ]
filter_log_expr = expr_values[intersect_gene, ]

#enrichment analysis
enrich_matrix <-
  get_enrich_matrix(filter_sig = filter_sig,
                    clustermarkers_list = clustermarkers_list)
enrich_result <-
  enrich_analysis(filter_log_expr = filter_log_expr,
                  enrich_matrix = enrich_matrix)
```

### 3.2.3 Spot deconvolution
Next, we applied DWLS deconvolution analysis to ST data to explore the proportion of scRNAseq cell type infiltrated in each spot.
```{r eval = FALSE}
DeconData <- SpatialDecon(
  enrich_matrix = enrich_matrix,
  enrich_result = enrich_result,
  filter_expr = filter_expr,
  filter_sig = filter_sig,
  clustermarkers_list = clustermarkers_list,
  meta_data = meta_data,
  malignant_cluster = "Malignant epithelial cells",
  tissue_cluster = "Epithelial cells",
  stromal_cluster = "Fibroblast cells"
  )
```

This step takes long time, in this vignette we read in our deconvoluted result.
```{r eval=FALSE}
DeconData <- openxlsx::read.xlsx(system.file("inst/extdata/DeconData.xlsx",package = "Cottrazm"))
```

```{r echo=FALSE}
DeconData <- openxlsx::read.xlsx('/work/xzz123/Project/2021/TumorBoundary/package/Cottrazm/inst/extdata/DeconData.xlsx')
print(head(DeconData))
```

### 3.2.4 Visualization
After getting the deconvolution result, we need to visualize the result with pie plot and bar plot. Pie plot was used to project the cell type infiltrated in each spot to spatial. Bar plot was used to compare cell type infiltrated in different locations including: malignant spots (Mal), boundary spots (Bdy), and non-malignant spots (nMal).
```{r eval=FALSE}
plot_col <- colnames(DeconData)[2:ncol(DeconData)]
img_path = system.file("inst/extdata/outs/spatial/tissue_lowres_image.png",package = "Cottrazm")

DeconPieplot(
  DeconData = DeconData,
  TumorST = TumorST,
  plot_col = plot_col,
  img_path = img_path,
  pie_scale = 0.4,
  scatterpie_alpha = 0.8,
  border_color = "grey"
  )

DeconBarplot(
  DeconData = DeconData,
  TumorST = TumorST,
  plot_col = plot_col
  )
```

```{r echo=FALSE}
plot_col <- colnames(DeconData)[2:ncol(DeconData)]
img_path = "/work/xzz123/Project/2021/TumorBoundary/package/Cottrazm/inst/extdata/outs/spatial/tissue_lowres_image.png"

  DeconData = DeconData;
  TumorST = TumorST;
  plot_col = plot_col;
  img_path = img_path;
  pie_scale = 0.4;
  scatterpie_alpha = 0.8;
  border_color = "grey"

DeconData <- DeconData[DeconData$cell_ID %in% rownames(TumorST@meta.data), ]

  ## Preprocess data
  slice <- names(TumorST@images)[1]

  spatial_coord <- data.frame(TumorST@images[[slice]]@coordinates) %>%
    tibble::rownames_to_column("cell_ID") %>%
    dplyr::mutate(
      imagerow_scaled =
        imagerow * TumorST@images[[slice]]@scale.factors$lowres,
      imagecol_scaled =
        imagecol * TumorST@images[[slice]]@scale.factors$lowres
    ) %>%
    dplyr::inner_join(DeconData, by = "cell_ID")

  ### Load histological image into R
  img_path <- img_path # lowers image png(input dir)
  img_frmt <- base::tolower(stringr::str_sub(img_path, -4, -1))

  if (img_frmt %in% c(".jpg", "jpeg")) {
    img <- jpeg::readJPEG(img_path)
  } else if (img_frmt == ".png") {
    img <- png::readPNG(img_path)
  }

  # Convert image to grob object
  img_grob <- grid::rasterGrob(img,
    interpolate = FALSE,
    width = grid::unit(1, "npc"),
    height = grid::unit(1, "npc")
  )


  ## Plot spatial scatterpie plot
  scatterpie_plt <- ggplot2::ggplot() +
    ggplot2::annotation_custom(
      grob = img_grob,
      xmin = 0,
      xmax = ncol(img),
      ymin = 0,
      ymax = -nrow(img)
    ) +
    scatterpie::geom_scatterpie(
      data = spatial_coord,
      ggplot2::aes(
        x = imagecol_scaled,
        y = imagerow_scaled
      ),
      cols = plot_col,
      color = border_color,
      alpha = scatterpie_alpha,
      pie_scale = pie_scale,
      lwd = 0.1
    ) +
    ggplot2::scale_y_reverse() +
    ggplot2::ylim(nrow(img), 0) +
    ggplot2::xlim(0, ncol(img)) +
    cowplot::theme_half_open(11, rel_small = 1) +
    ggplot2::theme_void() +
    ggplot2::coord_fixed(
      ratio = 1,
      xlim = NULL,
      ylim = NULL,
      expand = TRUE,
      clip = "on"
    )
  
 print(scatterpie_plt)
 
 ##Decon Bar
  DeconData = DeconData;
  TumorST = TumorST;
  plot_col = plot_col

 metadata <- DeconData
  metadata[, "Location"] <- TumorST@meta.data$Location[match(metadata$cell_ID, rownames(TumorST@meta.data))]

  Location <- do.call(rbind, lapply(unique(metadata$Location), function(x) {
    metadata_split <- metadata[metadata$Location == x, ] %>% as.data.frame()
    sub <- colSums(metadata_split[, plot_col], na.rm = T) %>%
      data.frame() %>%
      tibble::rownames_to_column() %>%
      set_colnames(., c("Types", "Sum")) %>%
      dplyr::mutate(Per = 100 * Sum / sum(Sum))
  }))
  Location$Group <- rep(unique(metadata$Location), each = length(plot_col))
  Location$Types <- factor(Location$Types, levels = plot_col)
  Barplot <- ggplot(Location, aes(x = Group, y = Per, fill = Types)) +
    geom_bar(stat = "identity") +
    theme_bw()
  print(Barplot)

```

## 3.3  Reconstruct of tumor spatial microencironment
Based on tumor boundary define result and spatial deconvolution result, we eager to reconstruct the tumor spatial microenrironment (TSME) at sub-spot level, which made up for the problem of insufficient resolution of ST data.

### 3.3.1 Reconstruction of spatial gene expression profile of scRNAseq cell types for sub-spots
Finally, we reconstructed the ST gene expression profile of scRNAseq cell types at sub-spot level based on ST data, scRNAseq data and spot deconvolution result, which could be used for further analysis. In this vignette, we only reconstruct the TSME of boundary spots, which was the region most likely associated with tumor progression, drug resistance and tumor metastasis. If you want to reconstruct TSME of whole ST data, you can change parameter Location = c("Mal","Byd","nMal"), but it will take longer time.
```{r eval=FALSE}
TumorSTRecon <-
  SpatialRecon(
    TumorST = TumorST,
    sig_exp = sig_exp,
    clustermarkers_list = clustermarkers_list,
    DeconData = DeconData,
    Location = "Bdy"
  )
```

### 3.3.2 Differentially expressed genes of tumor Boundary
To reconstruct of tumor spatial microenvironment, we then find differential expressed genes (DEG) in boundary spots:
```{r eval=FALSE}
DiffGenes <- FindDiffGenes(TumorST = TumorST, assay = "Spatial")
```

Nex, we visualized boundary DEG by volcano plot. If you want to visualize other types of ST data spots, you can change parameter Location to "Mal" or "nMal".
```{r eval=FALSE}
DiffVolcanoplot(
  DiffGenes = DiffGenes,
  Location = "Bdy",
  cut_off_pvalue = 2,
  cut_off_logFC = 0.25,
  n = 15
)
```

```{r echo=FALSE}
DiffGenes <- readr::read_rds('/work/xzz123/Project/2021/TumorBoundary/Fig/3.TransComposition/CRC1/1.SpatialDiffGenes.rds')
names(DiffGenes) <- c("nMal","Bdy","Mal")
  DiffGenes = DiffGenes;
  Location = "Bdy";
  cut_off_pvalue = 2;
  cut_off_logFC = 0.25;
  n = 15

# get Location DEG
  LocationDiff <- DiffGenes[[Location]]
  LocationDiff <- LocationDiff[LocationDiff$Symbol %in% grep("^IG[HJKL]|^RNA|^MT-|^RPS|^RPL", LocationDiff$Symbol, invert = T, value = T), ]

  dataset <- LocationDiff %>%
    tibble::rownames_to_column() %>%
    set_colnames(., c("gene", colnames(LocationDiff)))
  dataset$color <- ifelse(-log10(LocationDiff$FDR) < cut_off_pvalue | abs(LocationDiff$Diff) <= cut_off_logFC, "grey",
    ifelse(LocationDiff$Diff > 0, "Bdy", "Other")
  )
  datasetN <- dataset %>%
    dplyr::group_by(color) %>%
    dplyr::top_n(n, abs(Diff))
  datasetN <- datasetN[datasetN$color != "grey", ]

  # plot
  p <- ggplot(dataset, aes(x = Diff, y = (-log10(pvalue)), color = color)) +
    geom_point(aes(fill = color), size = 1) +
    scale_color_manual(values = c("red", "grey", "blue")) +
    ggrepel::geom_text_repel(
      data = datasetN, aes(label = datasetN$Symbol), max.overlaps = getOption("ggrepel.max.overlaps", default = 100),
      size = 3, point.padding = unit(0.35, "lines"), segment.color = "black", show.legend = F, color = "black", fontface = "bold"
    ) +
    geom_vline(xintercept = c(-cut_off_logFC, cut_off_logFC), lty = 4, col = "black", lwd = 0.8) +
    geom_hline(yintercept = cut_off_pvalue, lty = 4, col = "black", lwd = 0.8) +
    labs(x = "log2(FoldChange)", y = "-log10(pvalueue)") +
    theme(panel.background = element_rect(color = "black", fill = NA), legend.position = "bottom", legend.title = element_blank())
  print(p)
```

After find the DEG of boundary spots, we explored the KEGG and GO pathways theses features enriched in and speculated the possibly function of tumor boundary.
```{r eval=FALSE}
boundary_enrich <-
  FeatureEnrichment(
    DiffGenes = DiffGenes,
    cut_off_logFC = 0.25,
    cut_off_pvalue = 0.05,
    Location = "Bdy"
  )
```

## 4. Summary
Cottrazm includes three main function modules (BoundaryDefine, SpaitalDecon, and SpatialRecon) which allow users to easily analys ST tumor data. Here may still be gaps in the toolkit, but we'll work on more features later. If you have any suggestions or questions, please feel free to contact us (vera-morland@sjtu.edu.cn).

## 5. Session information
```{r}
sessionInfo()
```

## 6. Reference

